### PR TITLE
Add blocking and pubsub reserved connections

### PR DESCRIPTION
**Pull request by mbeasley**
Sidekiq enterprise and pro workers require the use of two categories of commands that redisbetween does not currently support:

- `subscribe`,`unsubscribe` (and `psubscribe`/`punsubscribe`) is used for live config changes to sidekiq workers (such as pausing work for a given a queue)
- `brpoplpush` is used as part of Sidekiq's reliable fetch mechanism to atomically pop and push jobs from queues to working queues in redis. The blocking part of this command is simply to add a timeout to prevent CPU saturation as Sidekiq polls the queue for new jobs to fetch.

Both of these commands differ from the current round trip style behavior of redisbetween, which is generally treated in a first come first serve manner. Both pubsub and blocking commands require connections in the redisbetween connection pool to be reserved, indefinitely if need be.

This feature branch implements two new classes of command handlers for redisbetween (versus the current round trip handler): `subscription` and `blocker`, both of which implement a `reservation` interface. Both implementations overlap a good bit, where incoming commands from local connections are grouped according to the channels they subscribe to or the source list that they are blocking on. Reservations still use connections from the main connection pool, but are limited via configuration. Reserved connections are returned to the pool when all local clients have unsubsribed or when the queue of blocking commands for a given `blocker` reservation has been emptied.

The primary difference between `subscription` and `blocker` is how messages are read from the upstream and then sent to the locals. `subscription` will subscribe to the appropriate channel and then listen for newly published messages, broadcasting any received message to all locals currently subscribed. `blocker` will form a queue of all incoming commands for the same source and send each command to the upstream in sequence, responding to the corresponding local once the upstream is unblocked or times out, and then moving to the next command in the queue.

`blocker`:
redisbetween_ blocker locking (2)

`subscription`:
redisbetween_ subscription locking

**Notes for Reviewing**
I recognize this format is a bit difficult for review when the discrete commit history is omitted. Sorry in advance with respect to that. Here are the primary concerns and areas to look at as a guide to the PR:
1. Implementation and tracking of reservations in `handlers/reservation.go` and across `proxy/proxy.go` and `handlers/command.go`.
2. Implementation of subscriptions (it's the simpler of the two), noting the subscribe/unsubscribe mechanisms, broadcasting, and thread safety.
3. Implementation of blockers, noting the enqueue/dequeue mechanisms, timeout adjustments, and thread safety.
4. Configuration constraints and enforcements in `config/config.go`.
5. Signal handling for quit/shutdown/kill.
6. Tests and the rest.

### CODE REVIEW

**Comment by mbeasley** [comment]
@dustMason - you may be interested in this given some of our past conversations.

**Comment by dustMason** [comment]
Hi 👋 thanks for the ping. Haven't had the time to look deeply yet, but do I understand correctly that for blocking / subscription commands, redisbetween will open and maintain a single connection to the upstream per client? eg, if multiple clients are all blocking on the same key via `BRPOPLPUSH`, redisbetween will maintain an upstream connection for each of them (up to the max)?

If that's the case, then is redisbetween performing any connection pooling? What's the added benefit of using redisbetween to proxy these blocking commands? Seems like we'd want a way to round-robin these blocking messages among connected clients while maintaining a single upstream connection. Likewise with subscriptions, we'd want to broadcast the same message to clients listening to the same key while holding one upstream connection. Apologies if this is already how it works and I'm misreading :)

File: handlers/blocker.go [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+}
+
+// Checks if command is a) singular and b) is a *blocking command
+func isBlockingCommand(commands []string) bool {
+	if len(commands) == 1 && strings.Contains(commands[0], "BRPOPLPUSH") {
+		return true
+	}
+
+	return false
+}
```
Comment (dustMason) [inline handlers/blocker.go:413]: I think this should also include

```
BLMOVE
BLMPOP
BLPOP
BRPOP
```

**Comment by mbeasley** [comment]
@dustMason not quite (thankfully). In both the subscription and blocking approaches, we group by the command in some way.

For subscriptions, we reserve a single upstream connection for any incoming local subscription commands that subscribe to the same channel (i.e. if we have 25 local connections all issuing `PSUBSCRIBE sidekiq-config`, we will group them all on the same upstream connection, keyed by `sidekiq-config`). We issue the subscribe once to the upstream and then broadcast published messages to all the locals that are tracked on the reservation. We return the reserved connection to the pool if all locals have unsubscribed.

Similarly for blocking commands, we reserve a single upstream connection for all local commands that share the same source argument of a `BRPOPLPUSH` command (i.e. if we have 25 local connections all issuing `BRPOPLPUSH default-queue working-queue 1`, we will group them all on the same upstream connection, keyed by `default-queue`). At this point, only `BRPOPLPUSH` is supported but adding others would not be too difficult. This implementation differs from subscriptions in that commands are tracked with their local connection in a queue on the reservation. We pop off the queue, send it to the upstream (letting it block as long as specified) and then round tripping the response to the corresponding local. We then move onto the next command in the queue, adjusting specified timeouts accordingly (for the amount of time already spent in the queue). We return the reserved connection to the pool if there are no more blocking commands in the queue.

File: handlers/blocker.go (hunk 1 above)
Comment (mbeasley) [reply handlers/blocker.go:413]: At this point, this is just adding `BRPOPLPUSH` but others wouldn’t be too difficult to add eventually.

**Review by mdehoog — APPROVED** [review]

**Comment by dustMason** [comment]
Ah, thanks for the explanation, that makes sense. Nice solution, well done 🙌

**Comment by cb-heimdall** [comment]
This Pull Request was merged without enough reviews
Please go [here](https://github.com/coinbase/redisbetween/issues/9) to review and +1

**Pull request merged**